common/src/main/java/com/wynntils/models/stats/StatCalculator.java

```diff
@@ -0,0 +1,197 @@
+            return RangedValue.of(lowerRollBound, higherRollBound);
+        } else {
+            int lowerRollBound = (int) Math.min(Math.ceil(lowerRawRollBound) - 1, 130);
+            int higherRollBound = (int) Math.max(Math.ceil(higherRawRollBound), 70);
+            return RangedValue.of(lowerRollBound, higherRollBound);
+        }
+    }
+
+    public static int calculateStarsFromInternalRoll(int internalRoll) {
+        // Star calculation reference, from salted:
+        // https://forums.wynncraft.com/threads/about-the-little-asterisks.147931/#post-1654183
+        int stars;
+        if (internalRoll < 101) {
+            stars = 0;
+        } else if (internalRoll < 125) {
+            stars = 1;
+        } else if (internalRoll < 130) {
+            stars = 2;
+        } else {
+            stars = 3;
+        }
+        return stars;
+    }
+
+    public static Pair<Integer, Integer> getDisplayRange(
+            StatPossibleValues possibleValues, boolean showBestValueLastAlways) {
+        StatType statType = possibleValues.statType();
+        RangedValue valueRange = possibleValues.range();
+        boolean isGood = valueRange.low() > 0;
+        Pair<Integer, Integer> displayRange;
+        int first;
+        int last;
+        if (showBestValueLastAlways || isGood) {
+            first = valueRange.low();
+            last = valueRange.high();
+        } else {
+            // Emulate Wynncraft behavior by showing the value closest to zero first
+            first = valueRange.high();
+            last = valueRange.low();
+        }
+        // We store "inverted" stats (spell costs) as positive numbers internally,
+        // but need to display them as negative numbers
+        if (statType.showAsInverted()) {
+            first = -first;
+            last = -last;
+        }
+        displayRange = Pair.of(first, last);
+        return displayRange;
+    }
+
+    public static float getPercentage(StatActualValue actualValue, StatPossibleValues possibleValues) {
+        int min = possibleValues.range().low();
+        int max = possibleValues.range().high();
+
+        float percentage = MathUtils.inverseLerp(min, max, actualValue.value()) * 100;
+        return percentage;
+    }
+
+    public static double getPerfectChance(StatPossibleValues possibleValues) {
+        // FIXME: This is the chance of getting a *** (3 star) roll, not the chance of
+        // getting the maximum possible value. But for now, keep old behavior.
+        double perfectChance = 1 / (possibleValues.baseValue() > 0 ? 101d : 61d) * 100;
+        return perfectChance;
+    }
+
+    public static double getDecreaseChance(StatActualValue actualValue, StatPossibleValues possibleValues) {
+        // If we are in any of these situations, our stat has "invalid" values. Maybe it is old?
+        // If so, rerolling will have a certain outcome of 100% or 0% chance of decrease
```

Comment:
An "old" roll can't give you an invalid value - the only time this would happen (bar Wynncraft mechanics fundamentally changing) is if our item API info is outdated. In that case, I don't think we should display any percentage, and it certainly shouldn't be 0 or 100

Reply:
So, the thing is, in the end, I did not rewrite the calculation code, I just took the old one and refactored it step by step to make it more sane (turn it into functions instead of a weird class which copied itself etc). I did not dare change the current behavior.

But my understanding here, which is reflected in the comment I wrote, is that if you have an item, which has an actual value outside the range of possible values, is either a bug or perhaps an old item where the Wynncraft gear type has changed but your specific item has not been updated (don't know if this can really happen).

But, I think the percentage is still correct. Note that you are looking at the "get decrease chance" function, not the basic "get percent of possible rolls". So if the range is 1-10, and you have a 13, then I claim that any reroll has 100% chance of getting lower, 0% chance of getting higher, and it would have a stat percentage of 130%. This is the best way we can describe this scenario if it happens, and this is what the code currently does.

If this scenario can ever happen, I don't know. But it's a simple piece of defensive programming to have this check.

Reply:
What you're describing is an impossible situation given the way Wynncraft currently works - so this code is defensive against a nonexistent scenario.

If a stat range is 1-10 and its actual value is 13, that doesn't mean that it's guaranteed to decrease after a reroll; it means our stat range is wrong, and we can't calculate anything about its reroll chances. To display any percentage at all is misleading, IMO

Reply:
Ok, I'll remove the defensive part.